tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/SetsOps.java

```diff
@@ -0,0 +1,141 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.framework.metrics.impl;
+
+import org.tensorflow.Operand;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.SparseOps;
+import org.tensorflow.op.sparse.DenseToDenseSetOperation;
+import org.tensorflow.types.family.TNumber;
+
+import static org.tensorflow.framework.utils.CastHelper.cast;
+
+/** Implementation of set operations */
+public class SetsOps {
+
+  /**
+   * Enumeration containing the string operation values to be passed to the TensorFlow Sparse Ops
+   * function {@link SparseOps#denseToDenseSetOperation}
+   */
+  public enum Operation {
+    A_MINUS_B("a-b"),
+    B_MINUS_A("b-a"),
+    INTERSECTION("intersection"),
+    UNION("union");
+
+    private final String setOperation;
+
+    Operation(String setOperation) {
+      this.setOperation = setOperation;
+    }
+
+    /**
+     * Gets the set operation String value used to pass as the stringOperation value to {@link
+     * SparseOps#denseToDenseSetOperation}
+     *
+     * @return the set operation String value
+     */
+    public String getSetOperation() {
+      return setOperation;
+    }
+  }
+
+  /**
+   * Computes set difference of elements in last dimension of <code>a</code> and <code>b</code> with
+   * <code>aMinusB</code> set to true/
+   *
+   * <p>All but the last dimension of <code>a</code> and <code>b</code> must match
+   *
+   * @param tf the TensorFlow Ops
+   * @param a The first operand representing set <code>a</code>
+   * @param b The other operand representing set <code>b</code>
+   * @param <T>the data type for the sets
+   * @return An Operand with the same rank as <code>a</code> and <code>b</code>, and all but the
+   *     last dimension the * same. Elements along the last dimension contain the results of the set
```

Comment:
This doesn't seem to accurately describe what is returned, given that we aren't returning a sparse vector.

For example, here is an except from the test code for this method (org/tensorflow/framework/metrics/impl/SetsOpsTest.java:71):
~~~ Java
        Operand<TInt32> a = tf.constant(new int[][] {{1, 5, 9}, {4, 5, 3}});
        Operand<TInt32> b = tf.constant(new int[][] {{1, 2, 6}, {1, 2, 2}});
        Integer[] expected = new Integer[] {5, 9, 3, 4, 5};
~~~
`expected` does not have the same rank as `a` and `b`. Nor does it have all but the last dimension the same.

Digging deeper into the call stack, the problem is that `setOperationResult.resultValues()` is returning the values of a sparse tensor without the indices or shape:

~~~ Java
  /**
   * 2D indices of a `SparseTensor`.
   */
  public Output<TInt64> resultIndices() {
    return resultIndices;
  }
  
  /**
   * 1D values of a `SparseTensor`.
   */
  public Output<T> resultValues() {
    return resultValues;
  }
  
  /**
   * 1D `Tensor` shape of a `SparseTensor`. `result_shape[0...n-1]` is
   * the same as the 1st `n-1` dimensions of `set1` and `set2`, `result_shape[n]`
   * is the max result set size across all `0...n-1` dimensions.
   */
  public Output<TInt64> resultShape() {
    return resultShape;
  }
~~~

I find myself again wondering if for now it would be simpler to use the element-wise version of (weightsShape = 1 or weightsShape = valuesShape)?

Reply:
OK I found the error, the output of` tf.sparse.denseToDenseSetOperation `needs to be run through
`tf.sparse.sparseToDense` to make a dense tensor.  Both operations have `indices`, `shape` and `values` but do not involve the native `SparseTensor` present in TF Python. The current set operations are limited to dense tensors.

Reply:
👍 done